xen/arch/arm/tee/ffa_rxtx.c

```diff
@@ -87,6 +101,69 @@ uint32_t ffa_handle_rxtx_map(uint32_t fid, register_t tx_addr,
     if ( !rx )
         goto err_unmap_tx;
 
+    /*
+     * Transmit the RX/TX buffer information to the SPM if acquire is supported
+     * as the spec says that if not there is not need to acquire/release/map
+     * rxtx buffers from the SPMC
+     */
+    if ( ffa_fw_support_fid(FFA_RX_ACQUIRE) )
+    {
+        struct ffa_endpoint_rxtx_descriptor_1_1 *rxtx_desc;
+        struct ffa_mem_region *mem_reg;
+        struct ffa_address_range *addr_rg;
+
+        /* All must fit in our TX buffer */
+        BUILD_BUG_ON((sizeof(*rxtx_desc) + sizeof(mem_reg)*2 +
+                      sizeof(addr_rg)*2) > FFA_PAGE_SIZE);
+
+        spin_lock(&ffa_tx_buffer_lock);
+        rxtx_desc = ffa_tx;
+
+        /*
+         * We have only one page for each so we pack everything:
+         * - rx region descriptor
+         * - rx region range
+         * - tx region descriptor
+         * - tx region range
+         */
+        rxtx_desc->sender_id = ffa_get_vm_id(d);
+        rxtx_desc->reserved = 0;
+        rxtx_desc->rx_region_offs = sizeof(*rxtx_desc);
+        rxtx_desc->tx_region_offs = sizeof(*rxtx_desc) + sizeof(mem_reg) +
+                                    sizeof(addr_rg);
+        /* rx buffer */
+        mem_reg = ffa_tx + sizeof(*rxtx_desc);
+        mem_reg->total_page_count = 1;
+        mem_reg->address_range_count = 1;
+        mem_reg->reserved = 0;
+
+        addr_rg = ffa_tx + sizeof(*rxtx_desc) + sizeof(mem_reg);
+        addr_rg->address = page_to_maddr(rx_pg);
+        addr_rg->page_count = 1;
+        addr_rg->reserved = 0;
+
+        /* tx buffer */
+        mem_reg = ffa_tx + sizeof(*rxtx_desc) + sizeof(mem_reg) +
+                  sizeof(addr_rg);
+        mem_reg->total_page_count = 1;
+        mem_reg->address_range_count = 1;
+        mem_reg->reserved = 0;
+
+        addr_rg = ffa_tx + sizeof(*rxtx_desc) + sizeof(mem_reg)*2 +
+                  sizeof(addr_rg);
+        addr_rg->address = page_to_maddr(tx_pg);
+        addr_rg->page_count = 1;
+        addr_rg->reserved = 0;
+
+        ret = ffa_rxtx_map(0, 0, 1);
+
+        spin_unlock(&ffa_tx_buffer_lock);
+
+        /* TODO: We could ignore errors and do not acquire/release after */
```

Comment:
from the spec the SPMC could just answer unsupported if it has no SP supporting indirect messages just because there is no case where it would need an access.